url7b90c1010b_DastanIqbal_HelloZMQwithJava_tgz-pJ8-ClientJ8/src/zmq/Router.java

```diff
@@ -0,0 +1,394 @@
+    
+    0MQ is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU Lesser General Public License for more details.
+        
+    You should have received a copy of the GNU Lesser General Public License
+    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+package zmq;
+
+import java.nio.ByteBuffer;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+//TODO: This class uses O(n) scheduling. Rewrite it to use O(1) algorithm.
+public class Router extends SocketBase {
+
+    public static class RouterSession extends SessionBase {
+        public RouterSession(IOThread io_thread_, boolean connect_,
+            SocketBase socket_, final Options options_,
+            final Address addr_) {
+            super(io_thread_, connect_, socket_, options_, addr_);
+        }
+    }
+    
+    //  Fair queueing object for inbound pipes.
+    private final FQ fq;
+
+    //  True iff there is a message held in the pre-fetch buffer.
+    private boolean prefetched;
+
+    //  If true, the receiver got the message part with
+    //  the peer's identity.
+    private boolean identity_sent;
+
+    //  Holds the prefetched identity.
+    private Msg prefetched_id;
+
+    //  Holds the prefetched message.
+    private Msg prefetched_msg;
+
+    //  If true, more incoming message parts are expected.
+    private boolean more_in;
+
+    class Outpipe
+    {
+        private Pipe pipe;
+        private boolean active;
+        
+        public Outpipe(Pipe pipe_, boolean active_) {
+            pipe = pipe_;
+            active = active_;
+        }
+    };
+
+    //  We keep a set of pipes that have not been identified yet.
+    private final Set <Pipe> anonymous_pipes;
+
+    //  Outbound pipes indexed by the peer IDs.
+    private final Map <Blob, Outpipe> outpipes;
+
+    //  The pipe we are currently writing to.
+    private Pipe current_out;
+
+    //  If true, more outgoing message parts are expected.
+    private boolean more_out;
+
+    //  Peer ID are generated. It's a simple increment and wrap-over
+    //  algorithm. This value is the next ID to use (if not used already).
+    private int next_peer_id;
+
+    // If true, report EHOSTUNREACH to the caller instead of silently dropping 
+    // the message targeting an unknown peer.
+    private boolean mandatory;
+
+    
+    public Router(Ctx parent_, int tid_, int sid_) {
+        super(parent_, tid_, sid_);
+        prefetched = false;
+        identity_sent = false;
+        more_in = false;
+        current_out = null;
+        more_out = false;
+        next_peer_id = Utils.generate_random (); 
+        mandatory = false;
+        
+        options.type = ZMQ.ZMQ_ROUTER;
+        
+        
+        fq = new FQ();
+        prefetched_id = new Msg();
+        prefetched_msg = new Msg();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
Analysis of this package determined that this line of code assigns a resource to a class field and there is no method that closes or returns this resource. This can result in a resource leak that slows down or crashes your system.

**Fix**
Create a method in this class that closes the following resource: *prefetched\_msg*. After you use this resource, use this method to prevent a resource leak.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).